### PR TITLE
Issue 22984 add new variants

### LINKED ISSUE #22984: [UI] Add Variants to experiments
## Is your feature request related to a problem? Please describe.
[Experiment] - Add a new variant

## Describe the solution you'd like
Make works the functionality of adding, editing, deleting, and viewing variants.

## Describe alternatives you've considered
N/A

## Acceptance criteria

- Add a new Variant, request a **required** name with a sidebar
Add variant name

- update the view
Create Experiments - Set Variants

## Additional context
Each added variation has these functionalities:
- View button (Only the primary)
- Edit button (Not the primary)
- Button to delete the variant (2 and 3)
- Create a sessionStorage with the name variantName of the variant clicked (View/Edit)
- Clickable weight to open the sidebar to change the Target split allocation (Not Yet implemented)

Screen Shot 2022-09-14 at 1 44 18 PM

- Update the tick to green

### PR DESCRIPTION
**Pull request by oidacra**
add/delete/view new variants.

### CODE REVIEW

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                                );
+                                this.closeSidebar();
+                            },
+                            (error: HttpErrorResponse) => throwError(error)
+                        )
+                    )
+                )
+            );
+        }
+    );
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts:92]: Effects belongs to the store, the component just call them.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    @Output()
+    closedSidebar = new EventEmitter<void>();
+
+    vm$ = this.componentStore.state$;
+    form: FormGroup<AddForm>;
+
+    private readonly addVariant = this.componentStore.effect(
+        (variant$: Observable<Pick<DotExperiment, 'name'>>) => {
+            return variant$.pipe(
+                tap(() => this.componentStore.patchState({ isSaving: true })),
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts:64]: Create an update for this and just call it.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        });
+    }
+
+    handleSubmit(): void {
+        const formValues = this.form.value as Pick<DotExperiment, 'name'>;
+        this.addVariant(formValues);
+    }
+
+    closeSidebar() {
+        this.componentStore.patchState({ isOpenSidebar: false });
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts:113]: Updaters are the ones that does this.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants/dot-experiments-configuration-variants.component.scss [hunk 4]
```diff
@@ -24,3 +18,29 @@ h2 {
...
+        border-top-right-radius: $border-radius-md;
+        border-top-left-radius: $border-radius-md;
+    }
+
+    &.last .p-card {
+        border-bottom-right-radius: $border-radius-md;
+        border-bottom-left-radius: $border-radius-md;
+        border: 1px solid $gray-lighter;
+    }
+}
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants/dot-experiments-configuration-variants.component.scss:46]: You can group all this inside a big `:host ::ng-deep` for all.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                        )
+                    )
+                )
+            );
+        }
+    );
+
+    constructor(
+        private readonly componentStore: ComponentStore<DotExperimentAddVariant>,
+        private readonly dotExperimentsConfigurationStore: DotExperimentsConfigurationStore,
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts:96]: Stores are one-to-one with a component. You should not have 2 stores in one component.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                            (error: HttpErrorResponse) => throwError(error)
+                        )
+                    )
+                )
+            );
+        }
+    );
+
+    constructor(
+        private readonly componentStore: ComponentStore<DotExperimentAddVariant>,
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts:95]: You shound not use ComponentStore directly in the component, you create a service and extend.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts (hunk 1 above)
Comment (oidacra) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants-add/dot-experiments-configuration-variants-add.component.ts:92]: moved.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants/dot-experiments-configuration-variants.component.ts [hunk 7]
```diff
@@ -16,28 +34,88 @@ import { DotIconModule } from '@dotcms/ui';
...
 
+    vm$ = this.dotExperimentsConfigurationStore.vmVariants$;
+
+    maxVariantsAllowed = MAX_VARIANTS_ALLOWED;
+    defaultVariantId = DEFAULT_VARIANT_ID;
+
+    private destroy$: Subject<boolean> = new Subject<boolean>();
+
+    constructor(
+        private readonly dotExperimentsConfigurationStore: DotExperimentsConfigurationStore,
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants/dot-experiments-configuration-variants.component.ts:60]: You should not be using this sore in two different components, also the store shuld be instantiated here in the component that uses it.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/store/dot-experiments-configuration-store.service.ts [hunk 8]
```diff
@@ -1,54 +1,82 @@
...
 import { HttpErrorResponse } from '@angular/common/http';
 import { DotExperimentsService } from '@portlets/dot-experiments/shared/services/dot-experiments.service';
 import { Title } from '@angular/platform-browser';
+import { DotMessageService } from '@services/dot-message/dot-messages.service';
+import { MessageService } from 'primeng/api';
 
 export interface DotExperimentsConfigurationState {
     pageId: string;
     experimentId: string;
     experiment: DotExperiment | null;
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/store/dot-experiments-configuration-store.service.ts:21]: Isn't `experimentId` the same as `experiment.id`?

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/dot-experiments-configuration.component.ts [hunk 9]
```diff
@@ -17,15 +13,18 @@ import { provideComponentStore } from '@ngrx/component-store';
 })
 export class DotExperimentsConfigurationComponent {
     pageId: string;
-    vm$: Observable<VmConfigurationExperiments> = this.dotExperimentsConfigurationStore.vm$.pipe(
-        tap((vm) => (this.pageId = vm.pageId))
-    );
+    vm$ = this.dotExperimentsConfigurationStore.vm$.pipe(tap((vm) => (this.pageId = vm.pageId)));
 
     constructor(
         private readonly dotExperimentsConfigurationStore: DotExperimentsConfigurationStore,
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/dot-experiments-configuration.component.ts:19]: Three.... :(

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/store/dot-experiments-configuration-store.service.ts [hunk 10]
```diff
@@ -93,7 +200,7 @@ export class DotExperimentsConfigurationStore
         });
     }
 
-    ngrxOnStoreInit() {
+    ngrxOnStateInit() {
         this.loadExperiment();
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/store/dot-experiments-configuration-store.service.ts:204]: I think there are 4 states changes only for init this the parent component.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/dot-experiments-configuration.component.html [hunk 11]
```diff
@@ -4,22 +4,17 @@
...
-            <p>
-                {{ vm.experiment.description }}
-            </p>
+            <p>{{ vm.experiment.description }}</p>
 
-            <dot-experiments-configuration-variants
-                [trafficProportion]="vm.experiment.trafficProportion"
-            >
-            </dot-experiments-configuration-variants>
+            <dot-experiments-configuration-variants></dot-experiments-configuration-variants>
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/dot-experiments-configuration.component.html:17]: Pass here from the store and receive here to update the store, don't use the store around.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants/dot-experiments-configuration-variants.component.ts [hunk 12]
```diff
@@ -16,28 +30,29 @@ import { DotIconModule } from '@dotcms/ui';
...
-        // to be implemented
-    }
-
-    addNewVariant() {
-        // to be implemented
-    }
+    @Input() status: StepStatus;
+    @Input() variants: Variant[];
+    @Output() showSidebar = new EventEmitter<void>();
+    @Output() hiddenSidebar = new EventEmitter<void>();
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants/dot-experiments-configuration-variants.component.ts:54]: Can't you unify this to one event that you can emit with data to know if open or close

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/dot-experiments-configuration.component.html [hunk 13]
```diff
@@ -4,22 +4,25 @@
...
 
             <dot-experiments-configuration-variants
-                [trafficProportion]="vm.experiment.trafficProportion"
-            >
-            </dot-experiments-configuration-variants>
+                [status]="vm.stepStatus"
+                [variants]="vm.experiment.trafficProportion.variants"
+                (delete)="deleteVariant($event)"
+                (goToEditPage)="goToEditPageVariant($event)"
+                (hiddenSidebar)="closeSidebar()"
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/dot-experiments-configuration.component.html:22]:
```suggestion
                (hideSidebar)="closeSidebar()"
```

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments-session-storage.service.ts [hunk 14]
```diff
@@ -0,0 +1,27 @@
+import { Injectable } from '@angular/core';
+import { SESSION_STORAGE_VARIATION_KEY } from '@portlets/dot-experiments/shared/models/dot-experiments-constants';
+
+@Injectable()
+export class DotExperimentsSessionStorageService {
```
Comment (fmontes) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/services/dot-experiments-session-storage.service.ts:5]: This looks like the perfect candidate for a global session storage service.

**Review by fmontes — DISMISSED** [review]

File: core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_button.scss [hunk 15]
```diff
@@ -365,6 +365,10 @@
     opacity: 1;
 }
 
+.p-button:disabled.p-button-link {
+    background-color: transparent !important;
+}
```
Comment (rjvelazco) [inline core-web/libs/dot-primeng-theme-styles/src/scss/dotcms-theme/components/_button.scss:370]: Doesn't it affect the UI/UX of other components?

**Review by rjvelazco — DISMISSED** [review]

**Comment by rjvelazco** [comment]
**This is a not code-related comment**

We have some files named `dot-${storeName}-store.service.ts` [E.g.: `dot-experiments-configuration-store.service.ts`]. 

In this case, we are adding the `.service` to a `store` file. In the `core-web` project we usually use `dot-${storeName}.store.ts`. 

I do not know which one is correct, `dot-${storeName}.store.ts` or `dot-${storeName}-store.service.ts` but we should set a convention. 

Not a big deal.

**Comment by oidacra** [comment]
> 

renamed.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants/dot-experiments-configuration-variants.component.spec.ts [hunk 16]
```diff
@@ -30,29 +47,181 @@ describe('DotExperimentsConfigurationVariantsComponent', () => {
...
+            );
+            expect(spectator.query(byTestId('variant-weight'))).toHaveText(
+                variantsVm.variants[0].weight + '.00% weight'
+            );
+            expect(spectator.query(byTestId('variant-preview-button'))).toExist();
+
+            expect(spectator.query(byTestId('variant-title-step-done'))).not.toHaveClass('isDone');
+        });
+
+        it('the variant(s)', () => {
```
Comment (rjvelazco) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-configuration/components/dot-experiments-configuration-variants/dot-experiments-configuration-variants.component.spec.ts:75]: Maybe?

```suggestion
        it('should load the variant(s)', () => {
```

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/components/dot-experiments-list-table/dot-experiments-list-table.component.scss [hunk 17]
```diff
@@ -68,7 +68,7 @@
                 tr {
                     height: 40px;
                     border: 1px solid $gray-lighter;
-                    border-width: 0 0 1px 0;
+                    //border-width: 0 0 1px 0;
```
Comment (rjvelazco) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/dot-experiments-list/components/dot-experiments-list-table/dot-experiments-list-table.component.scss:71]: Remove if it's not needed.

File: core-web/apps/dotcms-ui/src/app/view/components/_common/forms/dot-push-publish-form/dot-push-publish-form.component.spec.ts [hunk 18]
```diff
@@ -95,7 +95,7 @@ const mockPushActions: SelectItem[] = [
 
 const mockDate = new Date('2020, 8, 14');
 
-describe('DotPushPublishFormComponent', () => {
+xdescribe('DotPushPublishFormComponent', () => {
```
Comment (rjvelazco) [inline core-web/apps/dotcms-ui/src/app/view/components/_common/forms/dot-push-publish-form/dot-push-publish-form.component.spec.ts:98]: Yeah... These tests have been failing recently.

**Review by fmontes — APPROVED** [review]

**Review by rjvelazco — APPROVED** [review]

**Pull request merged**